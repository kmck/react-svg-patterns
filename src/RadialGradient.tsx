import * as React from 'react';

import {
  GradientColor,
  SpreadMethod,
  UnprocessedGradientStops,
} from './types';

import { SPREAD_METHOD_PAD } from './constants';
import { processStops } from './utils';

import GradientStop from './GradientStop';

interface Props {
  cx?: number,
  cy?: number,
  from?: GradientColor,
  fx?: number,
  fy?: number,
  id: string,
  r?: number,
  spreadMethod?: SpreadMethod,
  stops?: UnprocessedGradientStops,
  to?: GradientColor
};

const RadialGradient = ({
  cx = 0.5,
  cy = 0.5,
  from = 'inherit',
  fx = 0.5,
  fy = 0.5,
  id,
  r = 0.5,
  spreadMethod = SPREAD_METHOD_PAD,
  stops,
  to = 'inherit',
}: Props) => {
  const gradientStops = React.useMemo(
    () => processStops(stops || ([from, to])),
    [from, to, stops],
  );

  return (
    <radialGradient
      id={id}
      cx={cx}
      cy={cy}
      fx={fx}
      fy={fy}
      r={r}
      spreadMethod={spreadMethod}
    >
      {gradientStops.map((stopProps, key) => (
        <GradientStop
          key={stopProps.offset != null ? stopProps.offset : key}
          {...stopProps}
        />
      ))}
    </radialGradient>
  );
};

export default React.memo(RadialGradient);
