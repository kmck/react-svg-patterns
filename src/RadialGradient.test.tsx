import * as React from 'react';
import renderer from 'react-test-renderer';

import RadialGradient from './RadialGradient';

describe('LineraGradient', () => {
  it('renders a radial gradient', () => {
    const tree = renderer.create(
      <RadialGradient id="test" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a radial gradient using from/to', () => {
    const tree = renderer.create(
      <RadialGradient id="test" from="#ff0000" to="#0000ff" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a radial gradient using unprocessed stops', () => {
    const tree = renderer.create(
      <RadialGradient id="test" stops={['#ff0000', '#00ff00', '#0000ff']} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a radial gradient using preprocessed stops', () => {
    const tree = renderer.create(
      <RadialGradient id="test" stops={[{ offset: 0.25, color: '#000' }, { offset: 0.75, color: '#fff' }]} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a radial gradient with a radius', () => {
    const tree = renderer.create(
      <RadialGradient id="test" from="#fff" to="#000" r={2} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
    expect(tree).not.toEqual(renderer.create(
      <RadialGradient id="test" from="#fff" to="#000" />
    ).toJSON());
  });

  it('renders a radial gradient with a center offset', () => {
    const tree = renderer.create(
      <RadialGradient id="test" from="#fff" to="#000" cx={0.25} cy={0.75} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
    expect(tree).not.toEqual(renderer.create(
      <RadialGradient id="test" from="#fff" to="#000" />
    ).toJSON());
  });

  it('renders a radial gradient with a focus offset', () => {
    const tree = renderer.create(
      <RadialGradient id="test" from="#fff" to="#000" fx={0.25} fy={0.75} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
    expect(tree).not.toEqual(renderer.create(
      <RadialGradient id="test" from="#fff" to="#000" />
    ).toJSON());
  });

  it('renders a radial gradient with spreadMethod', () => {
    const tree = renderer.create(
      <>
        <RadialGradient id="test1" from="#fff" to="#000" spreadMethod="pad" />
        <RadialGradient id="test2" from="#fff" to="#000" spreadMethod="reflect" />
        <RadialGradient id="test3" from="#fff" to="#000" spreadMethod="repeat" />
      </>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
