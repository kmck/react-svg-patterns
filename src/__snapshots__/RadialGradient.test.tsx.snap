// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineraGradient renders a radial gradient 1`] = `
<radialGradient
  cx={0.5}
  cy={0.5}
  fx={0.5}
  fy={0.5}
  id="test"
  r={0.5}
  spreadMethod="pad"
>
  <stop
    offset="0%"
    stopColor="inherit"
  />
  <stop
    offset="100%"
    stopColor="inherit"
  />
</radialGradient>
`;

exports[`LineraGradient renders a radial gradient using from/to 1`] = `
<radialGradient
  cx={0.5}
  cy={0.5}
  fx={0.5}
  fy={0.5}
  id="test"
  r={0.5}
  spreadMethod="pad"
>
  <stop
    offset="0%"
    stopColor="#ff0000"
  />
  <stop
    offset="100%"
    stopColor="#0000ff"
  />
</radialGradient>
`;

exports[`LineraGradient renders a radial gradient using preprocessed stops 1`] = `
<radialGradient
  cx={0.5}
  cy={0.5}
  fx={0.5}
  fy={0.5}
  id="test"
  r={0.5}
  spreadMethod="pad"
>
  <stop
    offset="25%"
    stopColor="#000"
  />
  <stop
    offset="75%"
    stopColor="#fff"
  />
</radialGradient>
`;

exports[`LineraGradient renders a radial gradient using unprocessed stops 1`] = `
<radialGradient
  cx={0.5}
  cy={0.5}
  fx={0.5}
  fy={0.5}
  id="test"
  r={0.5}
  spreadMethod="pad"
>
  <stop
    offset="0%"
    stopColor="#ff0000"
  />
  <stop
    offset="50%"
    stopColor="#00ff00"
  />
  <stop
    offset="100%"
    stopColor="#0000ff"
  />
</radialGradient>
`;

exports[`LineraGradient renders a radial gradient with a center offset 1`] = `
<radialGradient
  cx={0.25}
  cy={0.75}
  fx={0.5}
  fy={0.5}
  id="test"
  r={0.5}
  spreadMethod="pad"
>
  <stop
    offset="0%"
    stopColor="#fff"
  />
  <stop
    offset="100%"
    stopColor="#000"
  />
</radialGradient>
`;

exports[`LineraGradient renders a radial gradient with a focus offset 1`] = `
<radialGradient
  cx={0.5}
  cy={0.5}
  fx={0.25}
  fy={0.75}
  id="test"
  r={0.5}
  spreadMethod="pad"
>
  <stop
    offset="0%"
    stopColor="#fff"
  />
  <stop
    offset="100%"
    stopColor="#000"
  />
</radialGradient>
`;

exports[`LineraGradient renders a radial gradient with a radius 1`] = `
<radialGradient
  cx={0.5}
  cy={0.5}
  fx={0.5}
  fy={0.5}
  id="test"
  r={2}
  spreadMethod="pad"
>
  <stop
    offset="0%"
    stopColor="#fff"
  />
  <stop
    offset="100%"
    stopColor="#000"
  />
</radialGradient>
`;

exports[`LineraGradient renders a radial gradient with spreadMethod 1`] = `
Array [
  <radialGradient
    cx={0.5}
    cy={0.5}
    fx={0.5}
    fy={0.5}
    id="test1"
    r={0.5}
    spreadMethod="pad"
  >
    <stop
      offset="0%"
      stopColor="#fff"
    />
    <stop
      offset="100%"
      stopColor="#000"
    />
  </radialGradient>,
  <radialGradient
    cx={0.5}
    cy={0.5}
    fx={0.5}
    fy={0.5}
    id="test2"
    r={0.5}
    spreadMethod="reflect"
  >
    <stop
      offset="0%"
      stopColor="#fff"
    />
    <stop
      offset="100%"
      stopColor="#000"
    />
  </radialGradient>,
  <radialGradient
    cx={0.5}
    cy={0.5}
    fx={0.5}
    fy={0.5}
    id="test3"
    r={0.5}
    spreadMethod="repeat"
  >
    <stop
      offset="0%"
      stopColor="#fff"
    />
    <stop
      offset="100%"
      stopColor="#000"
    />
  </radialGradient>,
]
`;
